### ISSUE TITLE
Robocasa demonstrations don't always succeed. 

### ISSUE THREAD

**Issue by velezbeltran** [robocasa/robocasa#209]
Hello!

Thank for this project it is very cool and super useful. 

I have been trying to verify that the Robocasa provided demonstrations work by replaying them in the simulator but it seems that this is unfortunately not the case. I was wondering how to fix them. 

It seems that there are two distinct issues. 

1. For some environments replaying the state in mujoco works. However, it seems that the check for success is not really correct. 
2. For some environments replaying the actions in the simulator doesn't lead to success (although replaying the states works so it seems to do more with the actions themselves). 

I choose 50 episodes at random for the 18 atomic tasks and ended up getting a success rate of 86%ish given these issues. 

It would also be helpful if you let us know what packages are used for the benchmarks or if there are any gotchas we should keep in mind.

Here are some more details I extracted with claude including our environment version and some things I tried as well as a minimal reproduction script. [further_details.md](https://github.com/user-attachments/files/29663434/GITHUB_ISSUE_DRAFT.1.md). 

Let me know if I can clarify or help with anything. 

Thank you in advance!

**Comment by Abhiram824** [comment]
Hi I will take a look at the state based replay failing to pass success check. 

However, unfortunately, open-loop action replay is very fragile is not expected to always result in a success.

**Comment by velezbeltran** [comment]
Thank you! Looking forward. And good to know :)

**Comment by kiwi142857** [comment]
We investigated the **state-replay** part and reproduced a fixture-pose / task-target synchronization issue affecting both `CloseBlenderLid` and `NavigateKitchen`.

Environment: RoboCasa `921c9a5`, robosuite `5ce6643`, MuJoCo `3.3.1`; headless, without policy inference.

The relevant reset sequence is:

1. `set_ep_meta(...)` followed by `env.reset()` reconstructs the Python fixture objects and samples their placements.
2. `reset_from_xml_string(...)` restores the recorded scene in MuJoCo.
3. Restoring the flattened state restores `time/qpos/qvel`, but the Python fixture poses and derived task targets can still correspond to the newly sampled scene.

For `CloseBlenderLid`, this mixes two sources of geometry: the current lid position comes from MuJoCo, whereas `get_lid_closed_pos()` uses the Python fixture pose through `get_pos_after_rel_offset()`, then caches the result. Clearing the cache alone does not help when its input pose is already inconsistent.

We tested this with **constructed successful states**, saved and restored through the replay sequence—not the original 50 demonstrations:

- **CloseBlenderLid:** three reconstructions shifted the checker target by 29.93, 12.43, and 28.17 cm. All changed from success to failure. Synchronizing the Python blender pose with the restored MuJoCo body pose and invalidating the cache restored success in each case.
- **NavigateKitchen:** two of four reconstructions shifted the navigation target by 30.51 and 26.19 cm, changing success to failure. Synchronizing fixture poses and recomputing the target restored success.

The restored flattened-state error was zero. The synchronization intervention did not change the simulator state or any success thresholds.

This suggests fixing the restoration contract: synchronize fixture geometry and invalidate/recompute derived targets after XML restoration, or explicitly serialize and restore task targets where appropriate. For the blender, computing the closed-lid target from the restored simulator body transform would avoid the stale Python-pose dependency.

Is there an intended synchronization hook that the playback path should call? We have not verified this against the original failing dataset episodes or performed a full regression test, and this finding does **not** establish the cause of the separate open-loop action-replay divergence.

**Comment by paulpacaud** [comment]
We also observe this issue when replaying ground-truth training actions in open-loop to sovle training episodes, and we observe far from perfect success rates:


Success Rates of training actions on training episodes
OpenDrawer | TurnOnMicroWave | PickPlaceCounterToStove | CloseBlenderLid | CoffeeSetupMug
96% | 94% | 94% | 70% | 91%

Do you recommend filtering out these failure training episodes for training policies?

**Comment by sattyamjjain** [comment]
This is the most load-bearing thread in robot evaluation right now, and it deserves a blunter framing: your benign arm is not at 100 percent, and almost nobody reports that.

Your ~86 percent over 50 replayed episodes is a benign floor. Any result of the form "perturbation X dropped success to 70 percent" measured on this suite is really "70 against a floor of 86", and the useful quantity is the difference, not the level.

The split you found is the important half. A wrong success check and a genuinely failing action replay are different bugs with opposite remedies, and separating them is what makes the floor reportable. Once you can say "of the 14 percent, N were scorer and M were replay", the scorer share is a fixed offset every downstream result can subtract, and the replay share is real physics that has to stay in.

Concretely, would you take a PR that emits, per task, the replay success rate and the scorer disagreement count in the same JSON as the evaluation output, so a harness can read the floor instead of assuming it?

**Comment by kiwi142857** [comment]
<html><head></head><body><p><span>Following up on our earlier comment: we have now validated the fixture-pose / task-target synchronization issue on </span><strong><span>900 released demonstrations</span></strong><span>, extending the previous constructed-state tests.</span></p><h3><span>Setup</span></h3><ul><li><p><span>RoboCasa </span><code><span>921c9a5</span></code><span>, robosuite </span><code><span>5ce6643</span></code><span>, MuJoCo </span><code><span>3.3.1</span></code><span>.</span></p></li><li><p><span>Official RoboCasa365 v1.0 </span><code><span>pretrain/atomic</span></code><span> demonstrations.</span></p></li><li><p><span>First 50 complete episodes per task, ordered by episode index, across 18 atomic-seen tasks.</span></p></li><li><p><span>State-based replay only, without policy inference or action execution.</span></p></li></ul><h3><span>Results</span></h3>
Task | Original checker | Geometry-synchronized checker
-- | -- | --
CloseBlenderLid | 4/50 | 50/50
NavigateKitchen | 39/50 | 50/50
Other 16 tasks | 800/800 | 800/800
Total | 843/900 (93.67%) | 900/900 (100%)

<p><span>All 57 changes were failure→success, with no regressions. All 900 released episodes had successful reward labels.</span></p><h3><span>Root cause and correction</span></h3><p><span>After restoring the recorded MuJoCo XML and state, Python-side fixture poses and derived targets can still describe the scene created during </span><code><span>env.reset()</span></code><span>.</span></p><ul><li><p><strong><span>CloseBlenderLid:</span></strong><span> we derived the closed-lid target from the restored MuJoCo blender body transform, avoiding the stale Python fixture pose/cache.</span></p></li><li><p><strong><span>NavigateKitchen:</span></strong><span> we recomputed the navigation target using fixture poses synchronized with the restored MuJoCo bodies.</span></p></li></ul><p><span>We preserved the original position, orientation, and gripper-distance thresholds. The correction did not change the recorded states, actions, masses, or reward labels.</span></p><p><span>This resolves the </span><strong><span>state-replay scorer disagreement on this sample</span></strong><span>. It does not resolve the separate open-loop action-replay divergence, and our deterministic first-50 selection is not necessarily the same sample used in the original report.</span></p><p><span>Would synchronizing fixture poses and invalidating/recomputing derived targets after XML/state restoration be the preferred upstream fix?</span></p></body></html>

**Comment by sattyamjjain** [comment]
@kiwi142857's 900-episode run answers the split I asked for, and the answer is cleaner than I expected: for state replay, it is 100 percent scorer, 0 percent physics. 843/900 to 900/900, all 57 changes from failure to success, no regressions. There is no residual floor to subtract once #225 lands. There is a bug to fix.

One correction to my own framing. I wrote "of the 14 percent, N were scorer, and M were replay" as though they would be mixed. The 86 percent in the original report is state and action replay combined; Kiwi's 93.67 is state only. Different denominators, and I should not have run them together.

The thing worth flagging, since nobody has answered it and two weeks have passed. @paulpacaud asked whether to filter failing episodes out of training sets. On these numbers, filtering against the pre-#225 checker would have been destructive. CloseBlenderLid scores 4/50 under the original checker and 50/50 under the fixed one. Anyone who dropped state-replay failures threw away about 92 percent of that task's demonstrations for a scorer bug rather than a data defect. It is also the task most likely to have been filtered, being the worst row in paulpacaud's action-replay table at 70 percent. Worth saying before more people act on the old numbers.

So the PR question I asked narrows. For state replay, after #225 the floor is zero, and there is nothing for a harness to read. For open-loop action replay, the floor is real and per-task, spanning 70 to 96 percent across paulpacaud's five. That is the number worth emitting beside evaluation output, and Abhiram824's "not expected to always succeed" makes it a permanent property rather than a bug to close.

#225 is 5 files, +558/-10, of which 421 lines are tests/test_state_restore.py. No review on it yet. @Abhiram824, you picked up the state-replay half in July; is this the shape you wanted?
